packages/pglite/src/base.ts

```diff
@@ -0,0 +1,328 @@
+   * templated statement where template values will be treated as parameters.
+   *
+   * You can use helpers from `/template` to further format the query with
+   * identifiers, raw SQL, and nested statements.
+   *
+   * This uses the "Extended Query" postgres wire protocol message.
+   *
+   * @param query The query to execute with parameters as template values
+   * @returns The result of the query
+   *
+   * @example
+   * ```ts
+   * const results = await db.sql`SELECT * FROM ${identifier`foo`} WHERE id = ${id}`
+   * ```
+   */
+  async sql<T>(
+    sqlStrings: TemplateStringsArray,
+    ...params: any[]
+  ): Promise<Results<T>> {
+    const { query, params: actualParams } = queryTemplate(sqlStrings, ...params)
+    return await this.query(query, actualParams)
+  }
+
+  /**
+   * Execute a SQL query, this can have multiple statements.
+   * This uses the "Simple Query" postgres wire protocol message.
+   * @param query The query to execute
+   * @returns The result of the query
+   */
+  async exec(query: string, options?: QueryOptions): Promise<Array<Results>> {
+    await this._checkReady()
+    // We wrap the public exec method in the transaction mutex to ensure that
+    // only one query can be executed at a time and not concurrently with a
+    // transaction.
+    return await this._runExclusiveTransaction(async () => {
+      return await this.#runExec(query, options)
+    })
+  }
+
+  /**
+   * Internal method to execute a query
+   * Not protected by the transaction mutex, so it can be used inside a transaction
+   * @param query The query to execute
+   * @param params Optional parameters for the query
+   * @returns The result of the query
+   */
+  async #runQuery<T>(
+    query: string,
+    params?: any[],
+    options?: QueryOptions,
+  ): Promise<Results<T>> {
+    return await this._runExclusiveQuery(async () => {
+      // We need to parse, bind and execute a query with parameters
+      this.#log('runQuery', query, params, options)
+      await this._handleBlob(options?.blob)
+      const parsedParams =
+        params?.map((p) => serializeType(p, options?.setAllTypes)) || []
+      let results
+      try {
+        results = [
+          ...(await this.#execProtocolNoSync(
+            serialize.parse({
+              text: query,
+              types: parsedParams.map(([, type]) => type),
+            }),
+            options,
+          )),
+          ...(await this.#execProtocolNoSync(
+            serialize.bind({
+              values: parsedParams.map(([val]) => val),
+            }),
+            options,
+          )),
+          ...(await this.#execProtocolNoSync(
+            serialize.describe({ type: 'P' }),
+            options,
+          )),
+          ...(await this.#execProtocolNoSync(serialize.execute({}), options)),
+        ]
+      } finally {
+        await this.#execProtocolNoSync(serialize.sync(), options)
+      }
+      this._cleanupBlob()
```

Comment:
nit: might want to `await` the cleanup for good measure, especially since for workers etc it might take a bit to happen and we don't want to release the lock before it happens?